// navalia/build/graphql/types/attr.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const graphql_1 = require("graphql");
const dom_types_1 = require("../dom-types");
exports.attr = {
    type: new graphql_1.GraphQLObjectType({
        name: 'attr',
        fields: () => ({
            value: {
                type: graphql_1.GraphQLString,
                description: `The value of the attribute you've queried for.`,
            },
        }),
    }),
    description: `The attr method operates similarly to the jQuery attr method, and return the value of an attribute of a DOM element. It’s called with 2 parameters: the css-style selector of the element you wish to query, and the attribute you want to retrieve.`,
    args: Object.assign({}, dom_types_1.domArgs, { selector: {
            description: `The selector you want to target (eg: '.buy-it-now')`,
            type: new graphql_1.GraphQLNonNull(graphql_1.GraphQLString),
        }, attribute: {
            description: `The attribute you want`,
            type: new graphql_1.GraphQLNonNull(graphql_1.GraphQLString),
        } }),
    resolve(_, args, context) {
        const { loader } = context;
        return loader.run(chrome => {
            return chrome
                .attr(args.selector, args.attribute, {
                wait: args.wait,
                timeout: args.timeout,
            })
                .then(value => {
                return { value };
            });
        });
    },
};
