// navalia/build/graphql/types/check.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const graphql_1 = require("graphql");
const dom_types_1 = require("../dom-types");
exports.check = {
    type: new graphql_1.GraphQLObjectType({
        name: 'check',
        fields: () => ({
            checked: {
                type: graphql_1.GraphQLBoolean,
            },
        }),
    }),
    description: `The check method checks a checkbox. It accepts a single argument: the css-style selector of the checkbox you want to check.`,
    args: Object.assign({}, dom_types_1.domArgs, { selector: {
            description: `The selector you want to target (eg: '.buy-it-now')`,
            type: new graphql_1.GraphQLNonNull(graphql_1.GraphQLString),
        } }),
    resolve(_, args, context) {
        const { loader } = context;
        return loader.run(chrome => {
            return chrome
                .check(args.selector, { wait: args.wait, timeout: args.timeout })
                .then(checked => {
                return { checked };
            });
        });
    },
};
