// navalia/build/graphql/types/click.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const graphql_1 = require("graphql");
const dom_types_1 = require("../dom-types");
exports.click = {
    type: new graphql_1.GraphQLObjectType({
        name: 'click',
        fields: () => ({
            clicked: {
                type: graphql_1.GraphQLBoolean,
                description: `Boolean representing if the click was successful`,
            },
        }),
    }),
    description: `The click method clicks on an element on the page. It accepts a css-style selector for the element you want to click. Returns a boolean indicating success`,
    args: Object.assign({}, dom_types_1.domArgs, { selector: {
            description: `The selector you want to target (eg: '.buy-it-now')`,
            type: new graphql_1.GraphQLNonNull(graphql_1.GraphQLString),
        } }),
    resolve(_, args, context) {
        const { loader } = context;
        return loader.run(chrome => {
            return chrome
                .click(args.selector, { wait: args.wait, timeout: args.timeout })
                .then(clicked => {
                return { clicked };
            });
        });
    },
};
