// navalia/build/graphql/types/cookie.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const graphql_1 = require("graphql");
exports.cookie = {
    type: new graphql_1.GraphQLList(new graphql_1.GraphQLObjectType({
        name: 'cookie',
        fields: () => ({
            name: {
                type: graphql_1.GraphQLString,
            },
            value: {
                type: graphql_1.GraphQLString,
            },
        }),
    })),
    description: `The check method checks a checkbox. It accepts a single argument: the css-style selector of the checkbox you want to check.`,
    args: {
        name: {
            description: `The name of the cookie you want to get or set`,
            type: graphql_1.GraphQLString,
        },
        value: {
            description: `The value of the cookie you want to set`,
            type: graphql_1.GraphQLString,
        },
    },
    resolve(_, args, context) {
        const { loader } = context;
        return loader.run(chrome => {
            return chrome.cookie(args.name, args.value);
        });
    },
};
