// navalia/build/graphql/types/coverage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const graphql_1 = require("graphql");
exports.coverage = {
    type: new graphql_1.GraphQLList(new graphql_1.GraphQLObjectType({
        name: 'coverage',
        fields: () => ({
            total: {
                type: graphql_1.GraphQLFloat,
            },
            unused: {
                type: graphql_1.GraphQLFloat,
            },
            percentUnused: {
                type: graphql_1.GraphQLFloat,
            },
        }),
    })),
    description: `The coverage method checks the coverage info for a particular resource (JS or CSS). In order to collect coverage, you must query goto with a second parameter of { coverage: true }. This is so that Navalia can properly instrument Chrome to collect coverage.`,
    args: {
        resource: {
            description: `The resource you want to check coverage on (CSS or JS)`,
            type: new graphql_1.GraphQLNonNull(graphql_1.GraphQLString),
        },
    },
    resolve(_, args, context) {
        const { loader } = context;
        return loader.run(chrome => {
            return chrome.coverage(args.resource).then(coverage => [coverage]);
        });
    },
};
