// navalia/build/graphql/types/exists.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const graphql_1 = require("graphql");
const dom_types_1 = require("../dom-types");
exports.exists = {
    type: new graphql_1.GraphQLObjectType({
        name: 'exists',
        fields: () => ({
            exists: {
                type: graphql_1.GraphQLBoolean,
                description: `If the selector was found on the page`,
            },
        }),
    }),
    description: `The exists method returns if the Element exists on the page. Accepts a string of a css-style selector.`,
    args: Object.assign({}, dom_types_1.domArgs, { selector: {
            description: `The selector you want to target (eg: '.buy-it-now')`,
            type: new graphql_1.GraphQLNonNull(graphql_1.GraphQLString),
        } }),
    resolve(_, args, context) {
        const { loader } = context;
        return loader.run(chrome => {
            return chrome
                .exists(args.selector, { wait: args.wait, timeout: args.timeout })
                .then(exists => {
                return { exists };
            });
        });
    },
};
