// navalia/build/graphql/types/focus.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const graphql_1 = require("graphql");
const dom_types_1 = require("../dom-types");
exports.focus = {
    type: new graphql_1.GraphQLObjectType({
        name: 'focus',
        fields: () => ({
            focused: {
                type: graphql_1.GraphQLBoolean,
                description: `Whether or not the selector was focused`,
            },
        }),
    }),
    description: `The focus method focuses an element on the page. It accepts a css-style selector of the element you wish to focus.`,
    args: Object.assign({}, dom_types_1.domArgs, { selector: {
            description: `The selector you want to target (eg: '.buy-it-now')`,
            type: new graphql_1.GraphQLNonNull(graphql_1.GraphQLString),
        } }),
    resolve(_, args, context) {
        const { loader } = context;
        return loader.run(chrome => {
            return chrome
                .focus(args.selector, { wait: args.wait, timeout: args.timeout })
                .then(focused => {
                return { focused };
            });
        });
    },
};
