// navalia/build/graphql/types/html.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const graphql_1 = require("graphql");
const dom_types_1 = require("../dom-types");
exports.html = {
    type: new graphql_1.GraphQLObjectType({
        name: 'html',
        fields: () => ({
            html: {
                type: graphql_1.GraphQLString,
                description: `The resulting HTML from the query`,
            },
        }),
    }),
    description: `The html method returns the string of HTML for a particular selector. It accepts one argument: a css-style selector for the element you wish to extract html from.`,
    args: Object.assign({}, dom_types_1.domArgs, { selector: {
            description: `The css-style selector you wish to query for`,
            type: graphql_1.GraphQLString,
        } }),
    resolve(_, args, context) {
        const { loader } = context;
        return loader.run(chrome => {
            return chrome
                .html(args.selector, { wait: args.wait, timeout: args.timeout })
                .then(html => {
                return { html };
            });
        });
    },
};
