// navalia/build/graphql/types/text.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const graphql_1 = require("graphql");
const dom_types_1 = require("../dom-types");
exports.text = {
    type: new graphql_1.GraphQLObjectType({
        name: `text`,
        fields: () => ({
            text: {
                type: graphql_1.GraphQLString,
                description: `The text inside of the selector`,
            },
        }),
    }),
    description: `The text method returns the inside-text of a DOM node (including its children’s text) as a string. It accepts one argument: a css-selector string, and defaults to body when none is present.`,
    args: Object.assign({}, dom_types_1.domArgs, { selector: {
            description: `The selector you want to get the text of`,
            type: new graphql_1.GraphQLNonNull(graphql_1.GraphQLString),
        } }),
    resolve(_, args, context) {
        const { loader } = context;
        return loader.run(chrome => {
            return chrome
                .text(args.selector, { wait: args.wait, timeout: args.timeout })
                .then(text => {
                return { text };
            });
        });
    },
};
