// navalia/build/graphql/types/type.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const graphql_1 = require("graphql");
const dom_types_1 = require("../dom-types");
exports.type = {
    type: new graphql_1.GraphQLObjectType({
        name: 'type',
        fields: () => ({
            typed: {
                type: graphql_1.GraphQLBoolean,
                description: `Whether or not the text was successfully typed.`,
            },
        }),
    }),
    description: `The type method allows you to type text into an element. It accepts two arguments: the css-style selector of the element you want to enter text into, and a string of text to input.`,
    args: Object.assign({}, dom_types_1.domArgs, { selector: {
            description: `The selector of the element you wish to type into`,
            type: new graphql_1.GraphQLNonNull(graphql_1.GraphQLString),
        }, text: {
            description: `The text you want to type into the element`,
            type: new graphql_1.GraphQLNonNull(graphql_1.GraphQLString),
        } }),
    resolve(_, args, context) {
        const { loader } = context;
        return loader.run(chrome => {
            return chrome
                .type(args.selector, args.text, {
                wait: args.wait,
                timeout: args.timeout,
            })
                .then(typed => {
                return { typed };
            });
        });
    },
};
