// navalia/build/graphql/types/uncheck.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const graphql_1 = require("graphql");
const dom_types_1 = require("../dom-types");
exports.uncheck = {
    type: new graphql_1.GraphQLObjectType({
        name: 'uncheck',
        fields: () => ({
            unchecked: {
                type: graphql_1.GraphQLBoolean,
                description: `Whether or not the uncheck was successful`,
            },
        }),
    }),
    description: `The uncheck method un-checks a checkbox. It accepts a single argument: the css-style selector of the checkbox you want to un-check`,
    args: Object.assign({}, dom_types_1.domArgs, { selector: {
            description: `The selector you want to target (eg: '.buy-it-now')`,
            type: new graphql_1.GraphQLNonNull(graphql_1.GraphQLString),
        } }),
    resolve(_, args, context) {
        const { loader } = context;
        return loader.run(chrome => {
            return chrome
                .uncheck(args.selector, { wait: args.wait, timeout: args.timeout })
                .then(unchecked => {
                return { unchecked };
            });
        });
    },
};
