// navalia/build/graphql/types/visible.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const graphql_1 = require("graphql");
const dom_types_1 = require("../dom-types");
exports.visible = {
    type: new graphql_1.GraphQLObjectType({
        name: 'visible',
        fields: () => ({
            visible: {
                type: graphql_1.GraphQLBoolean,
                description: `Whether or not the selector is visible.`,
            },
        }),
    }),
    description: `The visible method returns a boolean indiciating if an element is visible. It accepts a single argument: the css-style of the selector you want to check.`,
    args: Object.assign({}, dom_types_1.domArgs, { selector: {
            description: `The selector you want to target (eg: '.buy-it-now')`,
            type: new graphql_1.GraphQLNonNull(graphql_1.GraphQLString),
        } }),
    resolve(_, args, context) {
        const { loader } = context;
        return loader.run(chrome => {
            return chrome
                .visible(args.selector, { wait: args.wait, timeout: args.timeout })
                .then(visible => {
                return { visible };
            });
        });
    },
};
