// navalia/build/util/chrome.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const _ = require("lodash");
const axios_1 = require("axios");
const chromeLauncher = require("chrome-launcher");
const CDP = require("chrome-remote-interface");
exports.defaultFlags = {
    headless: true,
    disableGpu: true,
    hideScrollbars: true,
};
exports.transformChromeFlags = (flags) => {
    return _.chain(flags)
        .pickBy(value => value)
        .map((value, key) => {
        const cliSwitch = `--${_.kebabCase(key)}`;
        if (_.isBoolean(value) && value) {
            return cliSwitch;
        }
        return `${cliSwitch}=${value}`;
    })
        .value();
};
exports.launch = (flags, isHost = false, remote) => __awaiter(this, void 0, void 0, function* () {
    const logLevel = process.env.DEBUG &&
        (process.env.DEBUG.includes('ChromeLauncher') ||
            process.env.DEBUG.includes('*'))
        ? 'info'
        : 'silent';
    const chromeFlags = exports.transformChromeFlags(flags);
    const browser = remote
        ? Object.assign({}, remote, { kill: function () { }, pid: null })
        : yield chromeLauncher.launch({
            chromeFlags,
            logLevel,
        });
    const browserDebuggerURL = isHost
        ? yield axios_1.default
            .get(`ws://localhost:${browser.port}/json/version`)
            .then(res => res.data.webSocketDebuggerUrl)
        : '';
    const cdp = isHost
        ? yield CDP(remote || { target: browserDebuggerURL })
        : yield CDP(remote || { port: browser.port });
    yield Promise.all(isHost
        ? []
        : [
            cdp.Page.enable(),
            cdp.Runtime.enable(),
            cdp.Network.enable(),
            cdp.DOM.enable(),
            cdp.CSS.enable(),
        ]);
    return {
        browser,
        cdp,
    };
});
exports.createTab = (cdp, port) => __awaiter(this, void 0, void 0, function* () {
    const { browserContextId } = cdp.Target.createBrowserContext();
    const { targetId } = yield cdp.Target.createTarget({
        url: 'about:blank',
        browserContextId,
    });
    const tab = yield CDP({
        tab: `ws://localhost:${port}/devtools/page/${targetId}`,
    });
    yield Promise.all([
        tab.Page.enable(),
        tab.Runtime.enable(),
        tab.Network.enable(),
        tab.DOM.enable(),
        tab.CSS.enable(),
    ]);
    return {
        tab,
        targetId,
    };
});
