// navalia/build/util/ChromeHelper.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const debug = require("debug");
const chromeUtil = require("./chrome");
const Chrome_1 = require("../Chrome");
const log = debug('navalia:chrome-helper');
class ChromeHelper {
    constructor(options) {
        this.isExpired = false;
        this.jobsComplete = 0;
        this.activeTabs = 0;
        this.maxActiveTabs = options.maxActiveTabs || -1;
        this.flags = options.flags || chromeUtil.defaultFlags;
        this.timeout = options.timeout || 1000;
        log(`using up to ${this.maxActiveTabs === -1 ? 'infinite' : this.maxActiveTabs} tabs`);
    }
    start(callback) {
        this.activeTabs++;
        if (!this.browserStartingPromise) {
            log(`starting chrome`);
            this.browserStartingPromise = chromeUtil.launch(this.flags, true);
        }
        this.browserStartingPromise
            .then(launched => {
            this.kill = launched.browser.kill;
            this.port = launched.browser.port;
            this.cdp = launched.cdp;
            log(`chrome on ${this.port} is running`);
            return chromeUtil.createTab(this.cdp, this.port);
        })
            .then(({ tab, targetId }) => {
            log(`chrome on ${this.port} launched a new tab at ${targetId}. current tabs: ${this.activeTabs}`);
            const newTab = new Chrome_1.Chrome({
                cdp: tab,
                flags: this.flags || chromeUtil.defaultFlags,
                timeout: this.timeout,
            });
            newTab.on('done', this.onTabClose.bind(this, targetId));
            callback(newTab);
        });
    }
    onTabClose(targetId) {
        this.cdp.Target.closeTarget({ targetId });
        this.activeTabs--;
        this.jobsComplete++;
        log(`chrome on ${this.port} tab ${targetId} has closed. active tabs: ${this.activeTabs}, completed jobs: ${this.jobsComplete}`);
    }
    quit() {
        return __awaiter(this, void 0, void 0, function* () {
            log(`closing chrome ${this.port}, completed jobs: ${this.jobsComplete}`);
            this.activeTabs = 0;
            yield this.cdp.close();
            return this.kill();
        });
    }
    setExpired() {
        log(`chrome on ${this.port} has been set expired, active tabs: ${this.activeTabs}`);
        this.isExpired = true;
    }
    isIdle() {
        return this.activeTabs === 0;
    }
    isFull() {
        return this.maxActiveTabs === this.activeTabs || this.isExpired;
    }
    getIsExpired() {
        return this.isExpired;
    }
    getJobsComplete() {
        return this.jobsComplete;
    }
}
exports.ChromeHelper = ChromeHelper;
