// navalia/build/util/dom.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
function waitForElement(selector, timeout) {
    return new Promise((resolve, reject) => {
        const timeOutId = setTimeout(() => {
            reject(`Selector "${selector}" failed to appear in ${timeout} ms`);
        }, timeout);
        if (document.querySelector(selector))
            return resolve();
        const observer = new MutationObserver(function (_mutations, observation) {
            const found = document.querySelector(selector);
            if (found) {
                observation.disconnect();
                clearTimeout(timeOutId);
                return resolve();
            }
        });
        observer.observe(document, {
            childList: true,
            subtree: true,
        });
    });
}
exports.waitForElement = waitForElement;
function getPageURL() {
    return document.location.href;
}
exports.getPageURL = getPageURL;
function click(selector) {
    const element = document.querySelector(selector);
    if (!element) {
        throw new Error(`:click() > Unable to find element by selector: ${selector}`);
    }
    const event = document.createEvent('MouseEvent');
    event.initEvent('click', true, true);
    element.dispatchEvent(event);
    return true;
}
exports.click = click;
function html(selector) {
    const element = document.querySelector(selector);
    if (!element) {
        throw new Error(`:html() > Unable to find element by selector: ${selector}`);
    }
    return element.outerHTML;
}
exports.html = html;
