// navalia/build/graphql/ChromeLoader.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const Chrome_1 = require("../Chrome");
class ChromeLoader {
    constructor({ delay = 0 } = {}) {
        this.actions = [];
        this.timerId = 0;
        this.delay = 0;
        this.delay = delay;
        this.chrome = new Chrome_1.Chrome();
    }
    run(fn) {
        clearTimeout(this.timerId);
        return new Promise((resolve, reject) => {
            this.actions.push({
                handler: fn,
                resolve,
                reject,
            });
            this.timerId = setTimeout(this.serial.bind(this), this.delay);
        });
    }
    serial() {
        return this.actions.reduce((promise, action, idx, array) => {
            return promise.then(() => __awaiter(this, void 0, void 0, function* () {
                let res;
                try {
                    res = yield action.handler(this.chrome);
                    action.resolve(res);
                }
                catch (error) {
                    action.reject(error);
                }
                if (idx === array.length - 1)
                    this.chrome.done();
                return res;
            }));
        }, Promise.resolve());
    }
}
exports.ChromeLoader = ChromeLoader;
