// navalia/build/graphql/schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const graphql_1 = require("graphql");
const attr_1 = require("./types/attr");
const check_1 = require("./types/check");
const click_1 = require("./types/click");
const cookie_1 = require("./types/cookie");
const coverage_1 = require("./types/coverage");
const exists_1 = require("./types/exists");
const focus_1 = require("./types/focus");
const goto_1 = require("./types/goto");
const html_1 = require("./types/html");
const pageload_1 = require("./types/pageload");
const pdf_1 = require("./types/pdf");
const screenshot_1 = require("./types/screenshot");
const text_1 = require("./types/text");
const type_1 = require("./types/type");
const uncheck_1 = require("./types/uncheck");
const visible_1 = require("./types/visible");
const wait_1 = require("./types/wait");
exports.schema = new graphql_1.GraphQLSchema({
    query: new graphql_1.GraphQLObjectType({
        name: 'Browser',
        description: `All the various browser-related interactions can be queried here.`,
        fields: () => ({
            attr: attr_1.attr,
            check: check_1.check,
            click: click_1.click,
            cookie: cookie_1.cookie,
            coverage: coverage_1.coverage,
            exists: exists_1.exists,
            focus: focus_1.focus,
            goto: goto_1.goto,
            html: html_1.html,
            pageload: pageload_1.pageload,
            pdf: pdf_1.pdf,
            screenshot: screenshot_1.screenshot,
            text: text_1.text,
            type: type_1.type,
            uncheck: uncheck_1.uncheck,
            visible: visible_1.visible,
            wait: wait_1.wait,
        }),
    }),
});
