// navalia/build/Navalia.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const debug = require("debug");
const ChromeHelper_1 = require("./util/ChromeHelper");
const log = debug('navalia');
const isBusy = (chrome) => chrome.isFull();
const notBusy = (chrome) => !isBusy(chrome);
class Navalia {
    constructor(opts = {}) {
        this.numInstances = opts.numInstances || 1;
        this.maxJobs = opts.maxJobs || -1;
        this.workerTTL = opts.workerTTL || -1;
        this.chromeOptions = opts.chromeOptions || {};
        this.chromeInstances = [];
        this.queueList = [];
        log(`starting, using up to ${this.numInstances} instances`);
        for (let i = 0; i < this.numInstances; i++) {
            this.launchInstance(this.chromeOptions);
        }
        return;
    }
    destroy(chrome) {
        const instanceIndex = this.chromeInstances.indexOf(chrome);
        if (!chrome.isIdle()) {
            log(`instance ${chrome.port} still has active work, waiting till finished`);
            chrome.setExpired();
            return;
        }
        if (instanceIndex === -1) {
            return;
        }
        chrome.quit();
        this.chromeInstances.splice(instanceIndex, 1);
        this.launchInstance(this.chromeOptions);
        log(`instance ${chrome.port} successfully closed`);
    }
    execute(chromeHelper, queueItem) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!chromeHelper || (!queueItem || !queueItem.handler)) {
                throw new Error(`#execute was called with no instance of Chrome or a Job`);
            }
            chromeHelper.start((chrome) => __awaiter(this, void 0, void 0, function* () {
                log(`instance ${chromeHelper.port} is starting work`);
                try {
                    const result = yield queueItem.handler(chrome);
                    queueItem.resolve(result);
                }
                catch (error) {
                    queueItem.reject(error);
                }
                chrome.done();
                log(`instance ${chromeHelper.port} has completed work`);
                if (chromeHelper.isFull()) {
                    log(`instance ${chromeHelper.port} at max capacity, not taking work from queue`);
                    this.destroy(chromeHelper);
                    return;
                }
                if (chromeHelper.getIsExpired()) {
                    log(`instance ${chromeHelper.port} is expired and isn't taking new work`);
                    this.destroy(chromeHelper);
                    return;
                }
                if (chromeHelper.getJobsComplete() === this.maxJobs) {
                    log(`instance ${chromeHelper.port} has completed maximum jobs and is closing`);
                    this.destroy(chromeHelper);
                    return;
                }
                if (this.queueList.length) {
                    log(`instance ${chromeHelper.port} is taking work from the queue`);
                    return this.execute(chromeHelper, this.queueList.shift());
                }
            }));
        });
    }
    launchInstance(chromeOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const chromeHelper = new ChromeHelper_1.ChromeHelper(chromeOptions);
            this.chromeInstances.push(chromeHelper);
            if (this.queueList.length) {
                this.execute(chromeHelper, this.queueList.shift());
            }
            if (this.workerTTL > 0) {
                setTimeout(() => {
                    log(`instance ${chromeHelper.port} has reached expiration`);
                    this.destroy(chromeHelper);
                }, this.workerTTL);
            }
        });
    }
    kill() {
        return __awaiter(this, void 0, void 0, function* () {
            log(`:kill() > killing all instances regardless of work-in-progress`);
            return Promise.all(this.chromeInstances.map(chrome => chrome.quit()));
        });
    }
    run(handler) {
        return new Promise((resolve, reject) => {
            if (!this.chromeInstances.length || this.chromeInstances.every(isBusy)) {
                log('queueing work as all instances are busy');
                this.queueList.push({
                    handler,
                    resolve,
                    reject,
                });
                return;
            }
            log('instances are available and starting');
            this.execute(this.chromeInstances.find(notBusy), {
                handler,
                resolve,
                reject,
            });
        });
    }
}
exports.Navalia = Navalia;
